쥬스 메이커 [STEP 1] qoocrab, Whales

qoocrab opened the pull request:
안녕하세요 밈(@joSH0318)🐦!!
쥬스메이커 리뷰를 받게 된 `웨그랩` 조, qoocrab🦄과 Whales🐋입니다!!
3주동안 잘 부탁드립니다!!

### 고민한 점
> 효율적인 에러처리 방법

- `do-try-catch`는 로직이랑 예외처리가 혼재 되어있는 반면에 `Result`는 하나의 메서드 안에서 에러처리를 해주기 때문에 로직과 에러처리가 명확하게 분리된다고 생각해서 `Result`로 에러처리를 구현하였습니다.

> 재고 확인하는 방법
- 같은 쥬스를 한 번에 여러 잔 주문했을 때 재료의 재고를 확인하고 주문을 받을 수 있는지를 알려주는 canMakeJuice 함수를 구현했습니다. 확인하는 방법으로는 한 잔씩 여러 번 확인 할 지, 총 재료를 모두 합산하여 한 번에 확인 할 지 고민했었습니다.
- 곱하기 연산을 써서 한 번에 확인을 하는게 효율적이라고 생각해 후자의 방법을 선택했습니다.

> @main 누락 빌드 에러

- ![](https://hackmd.io/_uploads/Sk-OVoIEh.png)
사진과 같은 에러 문구와 함께 빌드에 자꾸 실패했습니다. 검색도 해보고 원인을 뒤져봤지만 찾지 못해서 결국 밈께 도움을 요청했습니다.
- 그 결과 중간에 실수로 `@main`을 지워버린 탓에 에러가 생겼다는 걸 알 수 있었습니다.
- `@main`은 프로그램 실행 시작 시 진입점으로 타입을 지정하기 위한 swift 언어의 기능으로, 사용자는 탑 레벨의 코드를 작성하는 대신 `@main` 단일 유형의 속성을 사용할 수 있고, 라이브러리와 프레임 워크는 프로토콜이나 클래스 상속을 통해 맞춤형 진입점 동작을 제공할 수 있습니다.
- 따라서 `Swift` 컴파일러가 `@main` 어노테이션을 통해 `AppDelegate`에서 전역범위에 있는 코드를 자동으로 인식하게 하고 실행시켜 `Entry Point`가 지정되게 하는데 이를 없애서 생겼던 빌드 에러였음을 공부할 수 있었습니다.

### 조언을 얻고 싶은 부분

> 에러처리 선택하기
- 로직이 복잡할 경우에는 `do-try-catch` 에러처리가 좋고, `Result`는 실행 속도가 `do-try-catch`에 비해 빠르다는 설명을 찾게 되었는데 어떤 부분에서 그런지 명확하게 이해가 어렵습니다.
- 언제 어느 것이 효율적일지 구별하기 위한 좋은 기준이 있을까요?

qoocrab committed:
feat: enum Fruit 생성, FruitStore - checkStock 함수 생성

- inventory를 만들어서 10개씩 재고 지정.
- checkStock 함수: 과일 인벤토리로부터 과일 개수를 확인
- checkStock dictionary 예외처리

WhalesJin committed:
feat: 과일 소모 함수 생성

과일과 양을 전달받아 inventory 재고를 변경하는 함수 생성

qoocrab committed:
feat: UseFruit 함수 생성

- UseFruit 함수: 쥬스를 만들기 위해 재고에서 과일을 꺼내는 함수
- JuiceError, Juice enum 생성

WhalesJin committed:
feat: 쥬스 만드는 함수 생성

switch 문을 이용하여 useFruits 함수 호출

qoocrab committed:
feat: orderFruit 함수 생성, Enum 파일 추가

- orderFruit: 과일 재고 추가하는 함수

WhalesJin committed:
feat: 주문 받는 함수 생성

- takeOrder : 쥬스 종류와 개수를 받아 메뉴를 만드는 함수 - Juice 타입에 쥬스이름 rawValue로 추가

qoocrab committed:
refactor: updateStock 함수 수정

- updateStock 함수로 하나로 합침

qoocrab committed:
refactor: makeJuice 리팩토링

- juiceRecipes를 만들고, makeJuice를 리팩토링함

WhalesJin committed:
feat: 주문량에 따른 재고 확인 함수 생성

WhalesJin committed:
refactor: 주문받는 함수 수정

파라미터 네이밍 수정, canMakeJuice 함수로 판매 가능 여부 확인

qoocrab committed:
fix: main 에러 수정

qoocrab committed:
refactor: stockUpdateResult 수정

qoocrab committed:
refactor: 컨벤션 수정

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMakerEnums.swift:
```diff
@@ -0,0 +1,32 @@
+
+enum StockUpdate {
+    case consume, order
+}
```
1. `StockUpdate`는 무슨 의미인가요? 사용되는 곳을 보면 `재고관련 작업 종류`라고 추측되는데 `StockUpdate`는 어색합니다.
2. `order`는 사용자가 Juice를 `order`하는 곳에서도 사용됩니다. 따라서 혼동의 여지가 있습니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMakerEnums.swift:
```diff
@@ -0,0 +1,32 @@
+    case insufficientFruitStock
+    case notFoundFruitInformation
+    case notFoundJuiceRecipe
+}
```
1. `FruitJuiceError`인데 `FruitStock`과 관련된 error가 있네요. 네이밍을 수정하면 좋겠습니다.
2. `notFoundFruitInformation`은 inventory에서 해당 과일을 찾지 못할 때 던져지는 error입니다. 때문에 네이밍이 충분히 그것을 설명하고 있지 않습니다.
3. 전체적으로 네이밍 수정이 필요해보여요! 🥹

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMakerEnums.swift:
```diff
@@ -0,0 +1,32 @@
+         strawNanaJuice = "딸바",
+         mangoJuice = "망고",
+         mangKiJuice = "망키"
+}
```
`rawValue` 사용하셨네요. 
`Juice.strawberryJuice.rawValue`와 같은 형태로 사용되겠네요.
사용하는 측면을 생각해본다면 저 `rawValue`가 의미하는 것을 알기 어렵습니다.
다른 방법을 사용해보면 좋겠어요.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,38 @@
+        .pineapple: 10,
+        .kiwi: 10,
+        .mango: 10
+    ]
```
`FruitStore`의 `inventory`의 `Fruit`들은 언제나 값이 `10`이 겠네요? 🧐
만약 `FruitStore` 인스턴스가 여러개 필요하다면 inventory의 과일 갯수는 언제나 `10`입니다.
확정성이 없는 코드로 보여집니다 🥹
다른 방법을 생각해볼까요?

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,38 @@
+        .mango: 10
+    ]
     
+    func checkStock(of fruit: Fruit) -> Int? {
```
`check`라는 동사는 반환값으로 Bool값을 예상하게 합니다.😄

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,38 @@
+    ]
     
+    func checkStock(of fruit: Fruit) -> Int? {
+        inventory[fruit]
```
return을 생략하셨군요. 
저의 개인적인 견해로는 가독성을 위해서 명시적으로 return을 써주는 것을 추천합니다.
또한, 다른 곳에서는 return을 생략하고 있지 않는 부분도 있어서 일관된 코드로 보여지지 않습니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,38 @@
+        }
+        
+        return .success(updateStock)
+    }
```
1. `updateStock`은 역할이 많습니다. 역할 분리를 하면 좋을 것 같아요.
2. `operation`을 parameter로 사용하셨는데 그 이유가 무엇인가요? 어떠한 이유가 있는지 궁금합니다.
3. `updateStock`에서 update는 동사로써 역할을 하는 것 같아요. 때문에 상수명으로 적합하지 않아보입니다!
4. `Result` 타입을 사용해보셨네요 👍 하지만 success인 경우 Int 타입의 `updateStock`을 반환하고 있는데 어디에 사용되고 있나요? 반환값으로 Int가 필요한가요? 만약 갱신된 재고 값이 필요하다면 필요한 시점에서 `checkStock`을 호출하면 될 것으로 보입니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,63 @@
+        .strawNanaJuice: [.strawberry: 10, .banana: 1],
+        .mangoJuice: [.mango: 3],
+        .mangKiJuice: [.mango: 2, .kiwi: 1]
+    ]
```
`juiceRecipes`를 이처럼 사용하면 범용성이 떨어질 것 같아요.
만약 다른 곳에서도 레시피를 사용하고 싶다면 어떨까요?
똑같은 코드를 반복해야하고, 그 과정에서 누락할 수 있는 가능성도 있어 보입니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,63 @@
+        if case .failure(let error) = stockUpdateResult {
+            throw error
+        }
+    }
```
1. `amount`는 레시피에 필요한 재료의 수를 뜻하는 것 같은데 네이밍 수정이 필요해보입니다. 이러한 애매한 네이밍은 가독성이 떨어지게 하는 요소가 됩니다.
2. 결국 반환된 Result에서 success는 사용되지 않겠네요? 그럼 왜 Result를 사용한 건지 궁금합니다. 또한, 어느 경우 Result 타입을 사용하는 것이 좋을지 생각해보시면 좋겠어요.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,63 @@
+    }
+    
+    private func makeJuice(_ juice: Juice) throws {
+        guard let recipe = juiceRecipes[juice] else {
```
어떠한 `recipe`인가요? 조금 더 명확한 네이밍을 부탁드립니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,63 @@
+        return true
+    }
+    
+    func takeOrder(_ juice: Juice, orderCount: Int) {
+        if canMakeJuice(juice, orderCount) {
+            for _ in 1...orderCount {
+                do {
+                    try makeJuice(juice)
+                } catch {
+                    print("Error: \(error)")
+                }
+            }
+            print("\(juice.rawValue)쥬스 \(orderCount)잔 나왔습니다.")
+        } else {
+            print("재고가 부족하여 주문을 받을 수 없습니다.")
+        }
+    }
```
1. 함수명에서 order라는 단어가 반복 사용되고 있습니다.
2. 이 method를 생각해보면, `쥬스의 종류와 주문 수를 받아 한번에 쥬스들을 만들고 있는 역할`로 보입니다.
하지만 여러가지 쥬스를 한번에 주문 받진 못하고 있네요? `한번에` 주문 받는 컨셉과 어긋나 보입니다.
일단 한가지 쥬스만 주문하는 method를 먼저 구현해보면 어떨까요?
3. rawValue가 의미하는게 무엇인지 파악하기 어렵네요.
4. `if - for - do `.. 3중 들여쓰기입니다. 🥹

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,63 @@
+        }
+        
+        for (fruit, recipeAmount) in recipes {
+            guard let stock = fruitStore.checkStock(of: fruit), stock >= recipeAmount * orderCount else {
```
guard문에 조건 두개를 이어서 사용하셨네요. 이런 경우 가독성이 떨어지는 단점이 생깁니다.
때문에 조건별로 개행을 주어서 가독성을 높여주는게 좋아보입니다.

JoSH0318 reviewed (commented):
안녕하세요 웨일, 쿠크랩 😀
첫번째 pr 고생하셨습니다. 
하지만 몇가지 아쉬운 점이 있네요 🥹

개인적으로 고민의 흔적이 적어보입니다. 더 나은 코드를 작성할 여지가 많은데 너무 아쉬운 부분입니다. 
확장성, 범용성, 유지보수 등을 생각한 코드, 네이밍이 부족해보여요.
또한, pr문을 통해서도 고민의 흔적 찾을 수 없어서 아쉬웠어요. 
`고민한 점`에서도 명확한 근거가 없어보이는 부분도 있고, `조언을 얻고 싶은 부분` 또한 무엇을 고민해서 어떠한 결론에 도달했고, 어떠한 질문을 하고싶은지 알수가 없었어요.

빠른 pr보단 조금 더 깊이 생각하는 기회를 가졌으면 좋겠어요.

---

### 질문 😀
1. `JuiceMaker`, `FruitStore`에 class와 struct를 사용하셨는데 각각 사용한 기준이 무엇인가요?
2. 여러분이 생각하는 `JuiceMaker`, `FruitStore`은 어떤 역할을 하는 객체일까요?
3. `고민한 점`에 Result 타입을 사용한 근거를 언급하셨습니다. 하지만 여기서 말하는 `효율적인`은 어떠한 측면에서의 효율인가요? 또한 `명확하게 분리`라고 말씀하셨는데 어떠한 부분이 명확하게 분리된다는 건지 이해하지 못했습니다. 조금더 구체적으로 말씀해주실수 있을까요?
4. `JuiceMakerEnums`파일에 모든 enum을 모아두셨는데, 그 기준이 무엇인가요? 파일과 타입 위치 또한 충분한 고민을 해보시면 좋을 것 같아요.
5. 이번 프로젝트에서 여러분이 생각하는 Model의 역할은 무엇인가요? 어디까지 인가요?
6. method 구현 순서 기준은 무엇인가요?

WhalesJin committed:
feat: 과일과 쥬스 열거형 타입 생성

FruitStore 타입을 struct로 수정, FruitStoreManager 타입 생성

WhalesJin committed:
feat: 과일 재고 늘리고 줄이는 함수 각각 생성

WhalesJin committed:
feat: 쥬스 레시피 타입 생성

쥬스를 키로, [과일: 개수]를 값으로 받는 Dictionary 생성, 레시피 추가,변경,삭제하는 method 생성

qoocrab committed:
feat: juice, fruit sturct를 생성

qoocrab committed:
feat: JuiceMaker struct 생성

WhalesJin committed:
feat: 쥬스 만드는 함수 생성

- 에러 타입 생성,
- 현재 재고량 받아오는 함수
- 재고량과 요구량 비교하는 함수
- 레시피북 제거

qoocrab committed:
feat: takeOrder 함수 생성

WhalesJin committed:
refactor: FruitStore 구조체를 클래스로 변경

- FruitStore method 순서 변경
- if문을 guard문으로 수정
- 쥬스 완성 중복 출력 수정

qoocrab committed:
fix: 로직 수정으로 인한 함수 네이밍 변경

WhalesJin commented:
### ﹖질문에 대한 답변
> 여러분이 생각하는 JuiceMaker, FruitStore은 어떤 역할을 하는 객체일까요?
- `JuiceMaker`는 쥬스 제조를 담당하는 객체이고, 고객의 주문을 받아 해당하는 쥬스를 만드는 역할을 합니다.특히, 필요한 과일의 종류와 수량을 파악하고, `FruitStore`에서 필요한 과일을 가져와 쥬스를 제조합니다. 만약 필요한 과일이 충분하지 않다면, 에러를 발생시킵니다.
- `FruitStore`는 과일 재고를 관리하는 객체이고, 과일의 재고를 증가시키거나 감소시키는 기능을 가지고 있습니다. 또한, 특정 과일의 현재 재고를 확인하는 기능도 제공합니다. `JuiceMaker`가 요청하는 과일의 제고를 제공하거나, 재고가 부족하다는 것을 알려주는 역할을 합니다.
> 고민한 점에 Result 타입을 사용한 근거를 언급하셨습니다. 하지만 여기서 말하는 효율적인은 어떠한 측면에서의 효율인가요? 또한 명확하게 분리라고 말씀하셨는데 어떠한 부분이 명확하게 분리된다는 건지 이해하지 못했습니다. 조금더 구체적으로 말씀해주실수 있을까요?
- `Result` 타입은 함수가 성공하거나 실패하였음을 명시적으로 표현하기 위해 사용됩니다. `Result` 타입은 두 가지 케이스를 가진 열거형으로, 하나는 `성공`, 하나는 `실패`와 관련된 오류를 가집니다.
- `Result` 타입은 처리 효율성 측면에서 효율적인데, 예로 성공과 실패 케이스를 명확하게 분리할 수 있습니다. 이는 함수의 반환 값을 처리할 때 효율성을 높여줍니다. 이는 코드를 더욱 단순하고 명확하게 만들어줍니다.
- 둘째로, 코드 `가독성` 측면에서 효율적입니다. 이는 코드를 읽는 사람에게 함수의 동작을 명확하게 이해시키는데 도움이 됩니다. 또한 오류 처리 코드를 함수 내부에서 분리하여 작성할 수 있습니다.
- 그러나 저희 프로젝트의 규모에서 성공과 실패를 구체적으로 분리하는 코드가 필요없다고 생각되어 차후 수정된 코드에서는 `Result` 타입을 사용한 부분들을 삭제하였습니다.
> JuiceMakerEnums파일에 모든 enum을 모아두셨는데, 그 기준이 무엇인가요? 파일과 타입 위치 또한 충분한 고민을 해보시면 좋을 것 같아요.
- 여러 가지 이유가 있을 수 있지만, 저희의 경우는 코드의 가독성 향상 측면에서 별도의 파일에 `enum` 타입을 모아두면 쉽게 확인할 수 있다는 장점이 있다고 생각햇습니다. 또한 중복 코드를 최소화해서 코드 유지 보수에 용이하기 위해라는 생각도 있었습니다. 그런데, 사실 특정 클래스나 구조체랑 연관이 강하게 있는 경우 같은 파일의 위치에 두는 것도 적합할 수 있겠다는 생각은 했습니다.
> 이번 프로젝트에서 여러분이 생각하는 Model의 역할은 무엇인가요? 어디까지 인가요?
- 저희가 학습한 MVC 모델에 따르면 모델은 데이터와 비즈니스 로직을 처리하는 역할을 담당하는 구성 요소입니다. 즉, 데이터 저장, 비즈니스 로직 처리, 데이터 접근 및 조작을 관리하는 역할이라고 생각합니다. 이후 스텝에서도 데이터와 비즈니스 로직의 구현이 필요한 부분은 모두 모델의 역할이라고 생각합니다.
> method 구현 순서 기준은 무엇인가요?
- 저희는 호출 순서에 따라 메소드를 배치하였습니다. `private` 메소드는 하단에 배치하는 것 또한 좋을 것이라는 생각도 들었지만, 본 프로젝트에서는 호출 순서를 고려하여 배치하였습니다.

### 📝 고민한 점
> 큰 그림이 중요하다!
- 사실 아직은 앱을 만드는 부분이나 UI부분이 낯설어서 step을 진행하며 전처럼 요구사항만 읽고 로직을 짜는데 집중했습니다. 그러다보니 진행하면서 넣을까 말까 고민하는 기능들, 넣고나니 더 꼬이기만 하는 기능들, 충돌되던 컨셉들을 발견하며 탁탁 막히는 느낌을 받았습니다.
- 프로젝트를 진행하며 `Main.storyboard`라는 부분을 보게 되었는데 이제까지 하던 고민이 결국 이로 이어지는 고민들이라는 것을 깨닫게 되었습니다. 이미 구성되어 있는 화면을 보며 타입 안에 필요한 프로퍼티나 메서드 등을 고민하고 구현 과정을 계획했다면 갈팡질팡 했던 긴 고민의 시간이 줄었을 것 같다는 생각을 했고, 앞으로 그런 방향도 같이 고려해서 로직을 짠다면 좀 더 효율적이고 완성도가 높은 코드를 구현하는데 큰 도움이 될 것 같다는 깨달음을 얻었습니다.
- (스토리보드를 미리 봤다면 고민의 시간이 줄었을 수 있는 부분)을 📱로 체크

> JuiceMaker, FruitStore에 class와 struct를 사용하셨는데 각각 사용한 기준이 무엇인가요?
- 처음에 코드를 새로 짤 때, 미션 파일에 `JuiceMaker`는 `struct`로, `FruitStore`는 `class`로 깔려있었는데 참조하거나 상속할 일이 없으니 class를 struct로 변경해서 진행했습니다.
- 완성 후 테스트를 돌려보는데
    ![](https://hackmd.io/_uploads/H1hMoEjN2.png)
    `qoocrab` 안에 `whales` 과일 가게가 있는데 `whales`에 있는 `fruitInventory`와 `qoocrab`에 있는 `fruitInventory`가 출력이 다르게 나오는 걸 보고 뭔가 잘못되었음을 깨달았습니다.
- 같은 재고를 가지고 있어야는데 `struct`를 써서 값 참조로 복사가 되었기 때문에 연동이 안됨을 찾아서 `FruitStore`를 `class`로 다시 바꿔서 해결했습니다.
- 프로젝트를 진행할 때마다 `class`와 `struct`에 대해서 좀 더 고민을 해보고 진행을 하면 좋겠다 라는 걸 직접 느끼게 되는 부분이었습니다.
- 📱

> 컨셉 잡기
- 클래스 (혹은 구조체)의 각 역할을 어디까지 부여하고 서로 어떤 방식으로 통신할지에 대해서 많은 고민을 했습니다. 처음에는 `StoreManager` 클래스를 두어서 주문을 받는 클래스를 따로 설정하고 여러가지 기능을 추가하고, `JuiceMaker`는 오로지 쥬스를 제조하기만 하는 역할로 두는 것을 생각했습니다. 그러나 차후 스텝들에서 프로그램의 목적이 쥬스를 제조하는 그 자체에 좀 더 초점이 있는 것 같아서 (버튼을 누르는 형태로 쥬스를 주문함) 이 두 가지 클래스를 분리하지 않고 합치는 것으로 합의를 보았습니다.
- 쥬스나 과일을 `struct`로 두어서 파일을 분리하는 것으로 합의를 보았습니다. 처음에는 쥬스나 과일에도 종류를 추가할 수 있는 확장성을 생각해서 이런 결정을 하였으나, 본 프로젝트의 의도는 정해진 과일과 쥬스 내에서 구현되는 것이라 이 부분은 수정을 하였습니다.
- `FruitStore`의 재고 관리를 `StoreManager`나 `JuiceMaker`에서 관리하는 것으로 처음에는 이야기가 되었으나, 이는 객체지향적으로 올바른 구현이 아니라고 생각이 되어 수정을 하였습니다.
- 📱

> JuiceRecipeBook
- `JuiceRecipeBook`이란 개념으로 쥬스 메이커가 이 레시피 북을 메모리에 관리하는 형태로 처음에 컨셉을 잡았었습니다. 그러나 이것이 메모리 낭비가 있고, 이후 테스트 코드를 작성할 시에 다양한 테스트를 하는 것을 막기 때문에 이 부분을 수정하는 것으로 합의하고 `JuiceRecipeBook`은 삭제 하였습니다.

> static
- 체인점의 개념을 생각했을 때, 모두의 레시피가 동일해야하고 한 군데의 레시피가 변경되면 다같이 바꾸기 위해서 static이라는 키워드를 떠올려보았습니다.
- `static` 변수는 여러 곳에서 공유하며 참조의 형태로 한 곳에서 변경되면 다같이 변경되는 특성을 가지고 있습니다.
- 단점은 프로그램이 종료될 때까지 메모리에 남아있어서 과도하게 사용하면 불필요한 메모리 소모가 발생될 수 있고, 여려 곳에서 공유하기 때문에 의도치 않게 변경될 수 있습니다.
- 추상화의 중요성을 느꼈습니다. 지금 우리가 구현해야하는 부분과 상관없는 고민임을 깨닫고 구현하지는 않았지만 덕분에 새로운 키워드를 공부하게 되었습니다.
- 📱

> Hashable 프로토콜 준수
- `Juice`라는 타입을 만들며 프로퍼티로 쥬스의 레시피를 `Dictionary`로 구현하였습니다. `Dictionary`가 `key-value` 값을 저장하는 해시 테이블로 구현이 되어있어서 그 `key`로 쓰여지는 타입은 `Hashable` 프로토콜을 준수해야 합니다. 그래서 `Fruit` 타입은 `Hashable` 프로토콜을 준수하도록 만들었습니다.
- 참고링크 : [블로그 - Hashable](https://babbab2.tistory.com/149)

> 확장성을 고려하는 시각 차이
- 확장성이라는 피드백을 받고 스토어를 관리하는 사장이 쥬스의 종류나 과일의 종류를 새롭게 추가할 수 있는 확장성을 생각하였습니다. 그에 따라 고려할 사항이 많아지면서 다양한 시도를 해보았습니다.
- 여기서의 확장성이란 기획자가 '메론쥬스 추가해줘'라고 했을 때 개발자가 다른 코드들은 크게 건드리지 않고 추가할 수 있는, 코드적인 측면에서의 확장성이라는 추가 설명을 듣고 코드들을 다시 수정하였습니다.
- 📱

> 네이밍
- 변수 네이밍 중 특히 `get`의 사용을 고민했다거나, 동사가 빠져있다거나 하는 경우가 많았었습니다. 그래서 이를 대체하는 동사를 고민을 많이 했습니다.
- `get`과 같은 예약어들은 사용처가 정해져 있기에 대체 단어가 있다면 피하는 게 좋고 make로 시작하는 `method`들은 보는 사람으로 하여금 `factory method`를 떠올리게 할 수 있어 피하는게 좋다.
- [API Design Guideline - Strive for Fluent Usage](https://www.swift.org/documentation/api-design-guidelines/#promote-clear-usage)

### 🙇🏻‍♀️ 조언을 얻고 싶은 부분

> 객체지향적인 사고
- 객체지향적인 구현이란 것이 무엇인지 정확하게 와닿지 않는 듯 합니다. 워낙 애매모호한 개념이지만 분명 일정 수준의 기준이 있다고는 생각하는데요. 저희가 얘기 나눈 부분에서 이러한 측면에서 의견 충돌(?)이 약간 있었고, 논의도 많이 있었습니다. 그래서 객체지향적 사고를 기를 수 있는 책이나 문서 등이 있으면 추천해주시면 감사하겠습니다. 그리고 저희가 다시 작성한 코드가 객체지향적으로 잘 구현이 되었는지에 대해서 구체적으로 평가를 해주시면 감사하겠습니다.

> 에러처리 구현
- 하나의 함수가 하나의 기능만 한다면 `takeOrder`함수는 `makeJuice`함수 호출 정도에서 그쳐야 한다는 생각을 들었습니다. 근데 그러면 에러처리는 새로 함수를 만들어야 하는지, 또 어디서 구현을 해야하는지에 대해 정리하지 못했습니다. 이 코드 그대로 괜찮은지 더 좋은 방향이 있는지 조언 부탁드립니다.

WhalesJin committed:
docs: README.md 작성

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by qoocrab, Whales on 2023/05/11.
+//
+
+enum Fruit: Hashable {
+    case strawberry, banana, pineapple, mango, kiwi
+}
```
`Hashable`을 빼도 오류가 생기지 않습니다.
왜 enum에는 `Hashable`을 채택하지 않아도 될까요?

[Hashable](https://developer.apple.com/documentation/swift/hashable) 공식문서를 읽어보면 enum에 채택시  설명이 나와있네요.😀 확인해보면 좋겠습니다.

qoocrab replied:
enum은 값 타입으로, enum이 고유한 해시 값을 가질 수 있음을 의미합니다. 따라서 enum이 기본적으로 Hashable을 준수해서 굳이 명시적으로 표시할 필요는 없습니다! 그러나 enum에 연관 값이 있는 경우에는 Hashable을 구현해야 합니다. 연관 값이 있는 enum은 각 케이스에 다른 값이 연결되어 있으므로 자동으로 해시 값을 계산할 수 없기 때문입니다~

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,37 @@
+        self.fruitInventory = fruitInventory
+    }
+    
+    func readCurrentStock(for fruit: Fruit) throws -> Int {
```
네이밍이 간결하고 뜻이 명확합니다. 좋습니다. 
다만 해당 method는 외부 접근을 허용한건가요? 은닉화가 안된 것 같아요. 🥹

WhalesJin replied:
핫 놓쳐서 다시 넣었습니다!! 감사합니다😃

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,37 @@
+    func isStockSufficient(_ fruit: Fruit, with requiredAmount: Int) -> Bool {
+        let currentStock = (try? readCurrentStock(for: fruit)) ?? 0
+        return currentStock >= requiredAmount
+    }
```
`readCurrentStock`에서는 로직이 실패할 경우 error를 던지도록 구현하셨어요.
하지만 `readCurrentStock`에서 error가 발생했을 때, `0`으로 바꿔주고 있네요.
애써 던진 error가 무의미해지고, `0`이라는 값이 유의미한 값이 아니라 아쉬운 코드네요. 🥹

WhalesJin replied:
`try`로 가져다 썼을 때 에러 처리 해야한다는 오류 문구가 나서 임시 수정을 했었는데 `throws`를 이용해서 에러 던지는 함수로 만들면 해결이 된다는 것을 찾았습니다!! 수정했습니다!!

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,37 @@
+    func decreaseFruitStock(_ fruit: Fruit, amount: Int) throws {
+        guard let currentAmount = fruitInventory[fruit] else {
+            throw JuiceMakerError.invalidFruit
+        }
```
`readCurrentStock`와 정확히 같네요. `readCurrentStock`을 호출하면 될 것 같아요!
이처럼 `readCurrentStock` method는 범용적으로 활용되면 좋을 것 같습니다.

WhalesJin replied:
오 주신대로 수정했더니 너무 좋네요!! 감사합니다!!!!!

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,37 @@
+    func increaseFruitStock(_ fruit: Fruit, amount: Int) throws {
+        guard let currentAmount = fruitInventory[fruit] else {
+            throw JuiceMakerError.invalidFruit
+        }
```
이 부분도 위의 언급과 같아요!

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMakerError.swift:
```diff
@@ -0,0 +1,10 @@
+//
+
+enum JuiceMakerError: Error {
+    case invalidFruit, invalidJuice
```
`invalidJuice`는 사용되는 곳이 없습니다.
이후에 사용할 것을 예상하고 만들어 놓은 것일까요?

WhalesJin replied:
넵넵 사용하게 되지 않을까요?ㅎㅎ

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,34 @@
+            if fruitStore.isStockSufficient(fruit, with: amount) {
+                try fruitStore.decreaseFruitStock(fruit, amount: amount)
+            } else {
+                try fruitStore.increaseFruitStock(fruit, amount: amount)
```
`makeJuice`는 메서드 이름으로 추측한다면,
쥬스를 만들고자 할 때 호출이 될 것으로 예상합니다.
그런데 `쥬스를 만든다 -> fruitStore에서 재고가 충분하면 재고 감소 or 충분하지 않으면 재고 추가` 형태로 동작하는게 어색합니다.

요구사항을 보시면 JuiceMaker 객체는
```
FruitStore의 과일을 사용해 과일쥬스를 제조합니다.
과일의 재고가 부족하면 과일쥬스를 제조할 수 없습니다.
```
라고 하고 있으니 재고 추가에 대한 동작은 현재 step에선 구현하지 않아도 될 것 같아요. 😀

WhalesJin replied:
넵넵!!! 그래서 에러 만들어서 에러로 처리했습니다!!!

JoSH0318 reviewed (commented):
고민한 점 / 조언을 얻고싶은 부분을 상세하게 적어주셔서 감사합니다.
이전에 너무 차갑게 코멘트를 한 것은 아닌지 생각이 듭니다. 🥹 죄송합니다. 🙇‍♂️

조언을 얻고 싶은 부분에 대한 답변과 코멘트 남겼습니다!

---

### 질문에 대한 답변

> Q: 객체지향적인 사고
객체지향적인 구현이란 것이 무엇인지 정확하게 와닿지 않는 듯 합니다. 워낙 애매모호한 개념이지만 분명 일정 수준의 기준이 있다고는 생각하는데요. 저희가 얘기 나눈 부분에서 이러한 측면에서 의견 충돌(?)이 약간 있었고, 논의도 많이 있었습니다. 그래서 객체지향적 사고를 기를 수 있는 책이나 문서 등이 있으면 추천해주시면 감사하겠습니다. 그리고 저희가 다시 작성한 코드가 객체지향적으로 잘 구현이 되었는지에 대해서 구체적으로 평가를 해주시면 감사하겠습니다.

A: 저도 객체지향이 아직 어려운 것 같아요. 현실세계의 객체를 코드의 객체로 설계한다는 것이 참 말처럼 쉽게 다가오는 개념은 아닌 것 같습니다. 하물며 사람마다 사물을 바라보는 관점도 다른데... 이것을 다시 코드의 객체로 구현한다는 것은 당연히 의견차이가 있을 수 밖에 없을 것 같아요!
서론이 너무 길었네요 ㅎㅎ 본론을 말하면 저는 `객체지향의 사실과 오해`라는 책을 읽었습니다. 저는 많은 사람이 읽는다길래 따라 읽은 케이스인데, 읽다보니 그렇게 어렵지도 않고 큰 도움이 됐던 것 같습니다. 너무 많은 시간을 투자하기 보단 가볍게 읽기 좋은 책 같습니다.

> Q: 에러처리 구현
하나의 함수가 하나의 기능만 한다면 takeOrder함수는 makeJuice함수 호출 정도에서 그쳐야 한다는 생각을 들었습니다. 근데 그러면 에러처리는 새로 함수를 만들어야 하는지, 또 어디서 구현을 해야하는지에 대해 정리하지 못했습니다. 이 코드 그대로 괜찮은지 더 좋은 방향이 있는지 조언 부탁드립니다.

A: "함수의 역할을 어디까지로 볼 것이냐"가 핵심 키워드인 것 같습니다. 
여러분이 말씀해주신 `하나의 함수는 하나의 기능만 한다.`에서 `하나의 기능`을 어디까지 볼 것이냐가 쟁점인 것 같아요.

`takeOrder`의 역할은 무엇일까에 대해서 생각했을 때, 웨일/쿠크랩은 `makeJuice`을 호출하는 것을 역할로 생각하신 것 같습니다.
만약 "`makeJuice`을 호출하고 그 결과로 error가 던져졌을 때, error를 처리하는 것"까지 역할로 생각한다면 어떨까요? `makeJuice`이 error를 던질 수도 있는 여지가 있는데 무책임하게 호출만 한다면 그것이 역할을 다하지 못했다고 생각이 들기도 합니다. 결과적으로 여러분의 코드는 그 책임을 다하고 있는 것 같습니다. 코드에 자신감을 가지셔도 될 것 같아요!

WhalesJin committed:
refactor: 피드백 받고 수정

- Hashable 제거
- 은닉화 수정
- FruitStore 에러 던지는 함수 수정
- makeJuice함수 재고 부족시 에러 던지는 형태로 수정

qoocrab commented:
@JoSH0318 
모두 반영하여 수정하였습니다! 확인 부탁드립니다.

qoocrab committed:
fix: consume이 쥬스 별로 일어나도록 함수 수정

WhalesJin committed:
feat: makeJuice 함수 생성

takeOrder 함수에 있던 if 문을 따로 makeJuice 함수로 생성

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,32 @@
+        return currentAmount
+    }
+    
+    func isStockSufficient(_ fruit: Fruit, with requiredAmount: Int) throws -> Bool {
+        let currentStock = try readCurrentStock(for: fruit)
+        return currentStock >= requiredAmount
+    }
```
Bool값을 return하는 것보다 `currentStock >= requiredAmount` 조건의 false인 경우  error를 throw하는 건 어떨까요.
`JucieMaker`의 `takeOrder`에서 오류 처리를 해주고 있으니 그래도 충분하다는 생각이 들어요.
이와 관련해서 이어지는 코멘트 `makeJuice`부분에 남기겠습니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,50 @@
+        .mango: 10
+    ])
+    
+    func takeOrder(_ juice: Juice) {
+        do {
+            try makeJuice(juice)
+        } catch JuiceMakerError.invalidFruit {
+            print("올바르지 않은 과일 이름입니다.")
+        } catch JuiceMakerError.invalidJuice {
+            print("올바르지 않은 쥬스 이름입니다.")
+        } catch JuiceMakerError.insufficientFruit {
+            print("과일의 재고가 부족하여 쥬스를 만들 수 없습니다.")
+        } catch {
+            print("알 수 없는 에러입니다.")
+        }
+    }
```
👍 가독성이 좋은 코드로 봅니다.
makeJuice 단계를 실행하고 오류가 생기면 catch하겠다는 흐름이 한눈에 보입니다.

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,50 @@
+        }
+    }
+    
+    func makeJuice(_ juice: Juice) throws {
+        if try canMake(juice) {
+            try consumeFruit(for: juice)
+            print("\(juice.name) 쥬스가 완성되었습니다.")
+        } else {
+            throw JuiceMakerError.insufficientFruit
+        }
+    }
+    
+    private func canMake(_ juice: Juice) throws -> Bool {
+        for (fruit, amount) in juice.recipe {
+            if !(try fruitStore.isStockSufficient(fruit, with: amount)) {
+                return false
+            }
+        }
+        return true
+    }
     
+    private func consumeFruit(for juice: Juice) throws {
+        for (fruit, amount) in juice.recipe {
+            try fruitStore.decreaseFruitStock(fruit, amount: amount)
+        }
+    }
```
해당 메서드들이 가독성이 떨어집니다.
예를 들어 
```swift
if !(try fruitStore.isStockSufficient(fruit, with: amount)) {
    return false
}
```
이 부분은 가독성이 떨어져요.
try를 보면 `fruitStore.isStockSufficient`가 error를 던질 수 있다는 걸 추론할 수 있어요. 
저는 개인적으로 `재고가 부족할 경우`도 error를 던져주겠구나 추측하게됐고, 때문에 `이것을 왜 Bool 값으로 return하여 위 메서드에서 error를 다시 만들어주고 있는거지?` 라는 의문이 들었습니다.

저라면 해당 코드를 아래와 같이 바꿔볼 것 같습니다.

```swift
func makeJuice(_ juice: Juice) throws {
    try verifyStock(juice)
    try consumeFruit(for: juice)
}

private func verifyStock(_ juice: Juice) throws {
     for (fruit, amount) in juice.recipe {
        try fruitStore.isStockSufficient(fruit, with: amount)
    }
}
    
private func consumeFruit(for juice: Juice) throws {
    for (fruit, amount) in juice.recipe {
        try fruitStore.decreaseFruitStock(fruit, amount: amount)
    }
}
```

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,50 @@
+        }
+    }
+    
+    func makeJuice(_ juice: Juice) throws {
+        if try canMake(juice) {
+            try consumeFruit(for: juice)
+            print("\(juice.name) 쥬스가 완성되었습니다.")
+        } else {
+            throw JuiceMakerError.insufficientFruit
+        }
+    }
```
`insufficientFruit` error는 FruiteStore에서 던져지는 게 맞는 것 같습니다.(재고가 부족하다는 에러니까요)
하지만 이 코드에선 `try canMake(juice)`가 아닐 경우, 해당 error를 던져주고 있습니다.
또한, `try canMake(juice)`는 이미 다른 error를 던지고 있습니다. 

코드를 좀 더 간결하게 만들 방법이 있는 것 같아요. 
`fruitStore.isStockSufficient`에서 재고가 부족할 경우 error를 던지고 다시 `canMake`에서 error를 던졌으면 합니다.
그러려면 위에서 언급한 것처럼 `isStockSufficient`의 return을 업애고 error를 던지도록 해야겠죠.

JoSH0318 reviewed (commented):
안녕하세요. 웨일, 쿠크랩!

몇가지 코멘트 남겼습니다. 😀
